browse/controllers/prevnext/__init__.py

```diff
@@ -0,0 +1,87 @@
+"""Handle requests to support sequential navigation between arXiv IDs."""
+
+from flask import url_for
+from typing import Tuple, Dict, Any
+from werkzeug import MultiDict
+from werkzeug.exceptions import InternalServerError, BadRequest
+
+from browse.domain.identifier import Identifier, IdentifierException
+from browse.services.database import get_sequential_id
+from arxiv import status
+from arxiv.taxonomy import ARCHIVES, CATEGORIES_ACTIVE
+from arxiv.base import logging
+
+
+Response = Tuple[Dict[str, Any], int, Dict[str, Any]]
+logger = logging.getLogger(__name__)
+
+
+def get_prevnext(request_params: MultiDict) -> Response:
+    """
+    Get the next or previous arXiv ID in the browse context.
+
+    The 'id', 'function', and 'context' request parameters are required. The
+    'site' parameter from the classic prevnext is no longer supported.
+
+    Parameters
+    ----------
+    request_params : dict
+
+    Returns
+    -------
+    dict
+        Search result response data.
+    int
+        HTTP status code.
+    dict
+        Headers to add to the response.
+
+    Raises
+    ------
+    InternalServerError
+        Raised when there was an unexpected problem executing the query.
+    BadRequest
+        Raised when request parameters are missing, invalid, or when an ID
+        redirect cannot be returned even when the request parameters are valid.
+
+    """
+    try:
+        if 'id' not in request_params:
+            raise BadRequest('Missing article identifier')
+        arxiv_id = Identifier(request_params['id'])
+
+        if not ('function' in request_params
+                and request_params['function'] in ['prev', 'next']):
+            raise BadRequest('Missing or invalid function request')
+
+        if 'context' not in request_params:
+            raise BadRequest('Missing context')
+        context = request_params['context']
+
+        if not (context in CATEGORIES_ACTIVE
+                or context in ARCHIVES or context == 'all'):
+            raise BadRequest('Invalid context')
+    except IdentifierException:
+        raise BadRequest(f"Invalid article identifier {request_params['id']}")
+
+    is_next = request_params['function'] == 'next'
+    try:
+        seq_id = get_sequential_id(paper_id=arxiv_id,
+                                   is_next=is_next,
+                                   context=context)
+        if seq_id:
+            redirect_url = url_for('browse.abstract',
+                                   arxiv_id=seq_id,
+                                   context=context)
+            return {},\
+                status.HTTP_301_MOVED_PERMANENTLY, {'Location': redirect_url}
+        else:
+            raise BadRequest(
```

Comment:
Should this be a 404 instead of a bad request?

Reply:
Probably, and was wondering during reimplementation (classic uses same). It seemed borderline enough that I didn't get bogged down by it.

Reply:
Sounds fine either way. 404 seems more correct with HTTP semantics since the same next request might work in the future.